GNUmakefile

```diff
@@ -1,26 +1,69 @@
-TEST?=$$(go list ./... |grep -v 'vendor')
+TEST?=./...
 GOFMT_FILES?=$$(find . -name '*.go' |grep -v vendor)
+PKG_NAME=pass
 
-default: build
+BINARY=terraform-provider-argocd
+VERSION = $(shell git describe --always)
 
-fmt:
-	gofmt -w $(GOFMT_FILES)
+default: build-all
+
+build-all: linux windows darwin
 
-build: fmt
+install: fmtcheck
 	go install
 
-test: fmt
-	go test -i $(TEST) || exit 1
-	echo $(TEST) | \
-		xargs -t -n4 go test $(TESTARGS) -timeout=30s -parallel=4
+linux: fmtcheck
+	@mkdir -p bin/
+	GOOS=linux GOARCH=amd64 go build -v -o bin/$(BINARY)_$(VERSION)_linux_amd64
+	GOOS=linux GOARCH=386 go build -v -o bin/$(BINARY)_$(VERSION)_linux_x86
+
+windows: fmtcheck
+	@mkdir -p bin/
+	GOOS=windows GOARCH=amd64 go build -v -o bin/$(BINARY)_$(VERSION)_windows_amd64
+	GOOS=windows GOARCH=386 go build -v -o bin/$(BINARY)_$(VERSION)_windows_x86
+
+darwin: fmtcheck
+	@mkdir -p bin/
+	GOOS=darwin GOARCH=amd64 go build -v -o bin/$(BINARY)_$(VERSION)_darwin_amd64
+	GOOS=darwin GOARCH=386 go build -v -o bin/$(BINARY)_$(VERSION)_darwin_x86
+
+release: clean linux windows darwin
+	for f in $(shell ls bin/); do zip bin/$${f}.zip bin/$${f}; done
+
+clean:
+	git clean -fXd -e \!vendor -e \!vendor/**/*
+
+test: fmtcheck
+	go test $(TEST) -timeout=30s -parallel=4
+
+testacc: fmtcheck
+	TF_ACC=1 go test $(TEST) -v -parallel 20 $(TESTARGS) -timeout 120m
+
+fmt:
+	@echo "==> Fixing source code with gofmt..."
+	gofmt -s -w ./$(PKG_NAME)
+
+# Currently required by tf-deploy compile
+fmtcheck:
+	@sh -c "'$(CURDIR)/scripts/gofmtcheck.sh'"
+
+lint:
+	@echo "==> Checking source code against linters..."
+	@GOGC=30 golangci-lint run ./$(PKG_NAME)
 
-testacc_prepare_env:
-	sh scripts/testacc_prepare_env.sh
+test-compile:
+	@if [ "$(TEST)" = "./..." ]; then \
+		echo "ERROR: Set TEST to a specific package. For example,"; \
+		echo "  make test-compile TEST=./$(PKG_NAME)"; \
+		exit 1; \
+	fi
+	go test -c $(TEST) $(TESTARGS)
 
-testacc:
-	sh scripts/testacc.sh
+vendor:
+	go mod tidy
+	go mod vendor
 
```

Comment:
These helper scripts make directives (testacc*) are useful to spin up a test cluster, at least I use them a lot :) can you leave them within the Makefile?